server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -507,7 +506,7 @@ public int hashCode() {
      * */
     class IndicesRequestCacheCleanupManager implements Closeable {
         private final Set<CleanupKey> keysToClean;
-        private final ConcurrentMap<ShardId, HashMap<String, Integer>> cleanupKeyToCountMap;
+        private final ConcurrentMap<ShardId, ConcurrentMap<String, Integer>> cleanupKeyToCountMap;
```

Comment:
Is it possible to refactor this as a `ConcurrentMap<Tuple<ShardId, String>, Integer>` (or similar) so that we don't have to reason about the concurrency properties of two maps operating together?

Reply:
This will take time, i can try to plan this for 2.16 however and dont want to miss the 2.15 deadline

Reply:
Here's the contract for all the `merge` and `compute` methods on concurrent map:

> When multiple threads attempt updates, map operations and the remapping function may be called multiple times.

Given that the remapping functions modify stateful things (the nested maps, and also other stateful class members), I'm having a hard time convincing myself the code here is correct.

Reply:
I don't quite understand your point, Are u suggesting to avoid merge and use compute as a whole to do the entire operation atomically ?

>I'm having a hard time convincing myself the code here is correct.

but i do have unit tests added to test the multithreaded scenario and i ran it for 1k times locally and there were no failures.

Reply:
@andrross We can't use a structure like` ConcurrentMap<Tuple<ShardId, String>, Integer>` or similar as we have a requirement to store a map of `<String, Integer>`(readerCacheKeyId and count) for a particular shardId. 
Also I didn't understand the last comment correctly. But the main problem earlier was using a non thread safe internal hashmap which was getting updated and accessed at the same time resulting into `concurrentModificationException`. As Kiran mentioned, we were able to reproduce this via UT and then fix it accordingly.

Reply:
> Also I didn't understand the last comment correctly.

In general you can't use a `ConcurrentMap` the way you're using it because you're assuming the remapping functions will only be invoked once.

I think maybe you _can_ use a ConcurrentHashMap this way as it has stronger guarantees, but you should change the type to be ConcurrentHashMap since you're explicitly relying on the behavior it provides that may not be offered by other ConcurrentMaps.

Reply:
Good catch. However it seems like `ConcurrentCollections.newConcurrentMap()` returns `ConcurrentHashMap` by default so we should be good for one. But definitely change in a follow-up PR to explicitly use `ConcurrentHashMap` here